### PR TITLE
fix(deps): Bump `@metamask/eth-json-rpc-middleware` to `^14.0.0`, `@metamask/transaction-controller` to `^35.1.1`

### LINKED ISSUE #26287: chore(deps): Bump `@metamask/eth-json-rpc-middleware` from `^12.1.1` to `^14.0.0`
### What is this about?

- Release diff: https://github.com/MetaMask/eth-json-rpc-middleware/compare/v12.1.1...v14.0.0
- Includes [major version bumps for dependencies](https://github.com/MetaMask/eth-json-rpc-middleware/compare/v12.1.1...v14.0.0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R31-R36) (`@metamask/{eth-block-tracker,eth-json-rpc-provider,json-rpc-engine,utils}`).
  - Fixes for breaking changes may be required to land this upgrade.
- Release blocker for confirmations.

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

_No response_

### Stakeholder review needed before the work gets merged

- [X] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

_No response_

### PR DESCRIPTION
**Pull request by MajorLift**
## **Description**

Updates `@metamask/eth-json-rpc-middleware` from `^12.1.1` to `^14.0.0`.
- This version bump comes with a large number of regressions, most of them type errors.
- This is because the package's dependencies are also updated by multiple major versions, and the changes include improved, stricter types (especially in `@metamask/utils`).

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/26143?quickstart=1)

## **Related issues**

- Closes https://github.com/MetaMask/metamask-extension/issues/26287
- Blocks:
  - https://github.com/MetaMask/MetaMask-planning/issues/2991
  - https://github.com/MetaMask/MetaMask-planning/issues/2810
  - https://github.com/MetaMask/metamask-extension/issues/25733

### Added

- Add and export `PPOMMiddlewareRequest` type for `JsonRpcRequest` types that include the `securityAlertResponse` property.
  - `securityAlertResponse` is defined as both optional and nullable.
- Add `PPOMRequest` type for `eth-sendTransaction` requests.

### Changed

- **BREAKING:** Bump `@metamask/eth-json-rpc-middleware` from `^12.1.1` to `^14.0.0`.
- **BREAKING:** Bump `@metamask/transaction-controller` from `^34.0.0` to `^35.1.1`.
- **BREAKING:** Redefine `SecurityAlertsAPIRequest` as a `JsonRpcRequest` type that accepts `unknown[]` as its `params` type.
- Widen the `request` parameters of the functions `validateWithController` and `validateWithAPI` to include `SecurityAlertsAPIRequest`.
- Bump `@trezor/connect-web` from `9.2.2` to `9.3.0`.

### Fixed

- **BREAKING:** Narrow `Params` generic parameter of `createPPOMMiddleware` function from `JsonRpcParams` to `(string | { to: string })[]`.
- Add `Params` generic parameter to `handleSnapRequest` function, which defaults to `JsonRpcParams`.
  - `handleSnapRequest` can now be typed correctly with any `params` object.

### Security

- **BREAKING:** Typed signature validation only replaces `0X` prefix with `0x`, and contract address normalization is removed for decimal and octal values.
  - Threat actors have been manipulating `eth_signTypedData_v4` fields to cause failures in blockaid's detectors.
  - Extension crashes with an error when performing Malicious permit with a non-0x prefixed integer address.
  - This fixes an issue where the key value row or petname component disappears if a signed address is prefixed by "0X" instead of "0x".

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by MajorLift** [comment]
Internal packages:

@SocketSecurity ignore npm/@lavamoat/lavapack@6.1.4
@SocketSecurity ignore npm/lavamoat-core@15.4.0
@SocketSecurity ignore npm/lavamoat-tofu@7.3.0
@SocketSecurity ignore npm/@metamask/snaps-utils@7.8.1
@SocketSecurity ignore npm/@metamask/eth-block-tracker@11.0.1

**Comment by MajorLift** [comment]
@SocketSecurity ignore npm/@storybook/node-logger@7.6.20
@SocketSecurity ignore npm/@storybook/client-logger@7.6.20
@SocketSecurity ignore npm/@storybook/postinstall@7.6.20
@SocketSecurity ignore npm/@storybook/react-dom-shim@7.6.20
@SocketSecurity ignore npm/@storybook/react@7.6.20
@SocketSecurity ignore npm/@storybook/addon-actions@7.6.20
@SocketSecurity ignore npm/@storybook/addon-docs@7.6.20
@SocketSecurity ignore npm/@storybook/core-common@7.6.20
@SocketSecurity ignore npm/@storybook/manager-api@7.6.20
@SocketSecurity ignore npm/@storybook/preview-api@7.6.20
@SocketSecurity ignore npm/@storybook/core-events@7.6.20
@SocketSecurity ignore npm/@storybook/blocks@7.6.20
@SocketSecurity ignore npm/@storybook/components@7.6.20
@SocketSecurity ignore npm/@storybook/csf-plugin@7.6.20
@SocketSecurity ignore npm/@storybook/csf-tools@7.6.20
@SocketSecurity ignore npm/@storybook/theming@7.6.20
@SocketSecurity ignore npm/@storybook/types@7.6.20
@SocketSecurity ignore npm/@storybook/channels@7.6.20
@SocketSecurity ignore npm/@storybook/docs-tools@7.6.20
@SocketSecurity ignore npm/@storybook/router@7.6.20
@SocketSecurity ignore npm/@storybook/core-client@7.6.20

Storybook is a trusted library.

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@SocketSecurity ignore npm/@storybook/addon-actions@7.6.20, npm/@storybook/addon-docs@7.6.20, npm/@storybook/blocks@7.6.20, npm/@storybook/channels@7.6.20, npm/@storybook/client-logger@7.6.20, npm/@storybook/components@7.6.20, npm/@storybook/core-client@7.6.20, npm/@storybook/core-common@7.6.20, npm/@storybook/core-events@7.6.20, npm/@storybook/csf-plugin@7.6.20, npm/@storybook/csf-tools@7.6.20, npm/@storybook/docs-tools@7.6.20, npm/@storybook/manager-api@7.6.20, npm/@storybook/node-logger@7.6.20, npm/@storybook/postinstall@7.6.20, npm/@storybook/preview-api@7.6.20, npm/@storybook/react-dom-shim@7.6.20, npm/@storybook/react@7.6.20, npm/@storybook/router@7.6.20, npm/@storybook/theming@7.6.20, npm/@storybook/types@7.6.20

**Comment by MajorLift** [comment]
@SocketSecurity ignore npm/@metamask/eth-block-tracker@11.0.1, npm/lavamoat-core@15.4.0, npm/lavamoat-tofu@7.3.0

Internal packages

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@SocketSecurity ignore npm/diff@5.2.0 
benign author change

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@metamaskbot update-policies

File: ui/store/actions.ts
Comment (MajorLift) [inline ui/store/actions.ts:1248]: Since the `request` argument is polymorphic, its generic parameter(s) should be exposed in the outer function as well.

The only exception to this would be if we want to force downstream consumers to use a certain fixed generic parameter by including it in closure.

In this case, `Params` is required to correctly type [`rawSnapRequestFunction`](https://github.com/MetaMask/metamask-extension/pull/26143/files#diff-72c459ae2e3f5674fd06869a208f0600cf4027dd04a5b1638db1677949010120R98) and ['uploadFile'](https://github.com/MetaMask/metamask-extension/pull/26143/files#diff-72c459ae2e3f5674fd06869a208f0600cf4027dd04a5b1638db1677949010120R194) in the `snap-interface.tsx` file.

File: app/scripts/lib/ppom/security-alerts-api.ts
Comment (MajorLift) [inline app/scripts/lib/ppom/security-alerts-api.ts:16]: Defines expected type for `JsonRpcRequest` objects that are compatible with the security alerts API.

File: app/scripts/lib/ppom/ppom-util.ts
Comment (MajorLift) [inline app/scripts/lib/ppom/ppom-util.ts:209]: In [`validateRequestWithPPOM`](https://github.com/MetaMask/metamask-extension/pull/26143/files#diff-56c892dd2b3324d3e12096b6e2280dafd007572a2c6797f8f6edc9a6f782b293R56) the output from `normalizePPOMRequest` is passed in as the `request` argument. 

Because the return type of `normalizePPOMRequest` includes `SecurityAlertsApiRequest`, the type of the request parameter here also needs to be widened.

File: app/scripts/lib/ppom/ppom-util.ts
Comment (MajorLift) [inline app/scripts/lib/ppom/ppom-util.ts:224]: Same as above, the output from `normalizePPOMRequest` is passed in as the `request` argument in [`validateRequestWithPPOM`](https://github.com/MetaMask/metamask-extension/pull/26143/files#diff-56c892dd2b3324d3e12096b6e2280dafd007572a2c6797f8f6edc9a6f782b293R55).

File: app/scripts/lib/ppom/ppom-util.ts
Comment (MajorLift) [inline app/scripts/lib/ppom/ppom-util.ts:144]: Narrows `request` to `JsonRpcRequest`.

File: app/scripts/lib/ppom/ppom-util.ts
Comment (MajorLift) [inline app/scripts/lib/ppom/ppom-util.ts:148]: `transactionParams` is narrowed to `PPOMRequest` due to the type guard above.

File: app/scripts/lib/ppom/ppom-util.ts
Comment (MajorLift) [inline app/scripts/lib/ppom/ppom-util.ts:39]: Define expected type for `eth-sendTransaction` requests.

File: app/scripts/lib/ppom/ppom-middleware.ts
Comment (MajorLift) [inline app/scripts/lib/ppom/ppom-middleware.ts:34]: Defined `securityAlertResponse` property as both optional and nullable, since both patterns are used in the tests.

File: app/scripts/lib/ppom/ppom-middleware.ts
Comment (MajorLift) [inline app/scripts/lib/ppom/ppom-middleware.ts:55]: Narrowed generic constraint based on the expected shape of the request params object outlined in the function body.

File: app/scripts/lib/createDupeReqFilterStream.test.ts
Comment (MajorLift) [inline app/scripts/lib/createDupeReqFilterStream.test.ts:29]: Mock requests with no `id` property are used in this test file, which are `JsonRpcNotification` by definition.

File: .yarn/patches/@trezor-connect-web-npm-9.3.0-040ab10d9a.patch
Comment (MajorLift) [inline .yarn/patches/@trezor-connect-web-npm-9.3.0-040ab10d9a.patch:1]: No user-facing changes. Applied analogous changes found in previous patch version.

**Review by MajorLift — COMMENTED** [review]
Notes for reviewers:

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@metamaskbot update-policies

File: ui/contexts/snaps/snap-interface.tsx
Comment (Mrtenz) [inline ui/contexts/snaps/snap-interface.tsx:114]: Are these guaranteed to never be `null` now?

**Comment by bschorchit** [comment]
Hey, is there anything we can do to support/speed up this going in. It contains a fix for a bug that crashes the extension https://github.com/MetaMask/metamask-extension/issues/25733.

**Comment by MajorLift** [comment]
@metamaskbot update-policies

File: ui/contexts/snaps/snap-interface.tsx
Comment (MajorLift) [reply ui/contexts/snaps/snap-interface.tsx]: I assumed that since these are json-rpc-request params `null` would be a valid type, but if not I can restore the previous code. 

Reverted here: e0f9925915cef9e62512ea5d892951b4fa685662

**Comment by MajorLift** [comment]
@metamaskbot update-policies

**Comment by MajorLift** [comment]
@legobeat I bumped `@metamask/transaction-controller` to latest (https://github.com/MetaMask/metamask-extension/pull/26143/commits/3c409340e3f46f385d1f8b18c3be5af32b2fe631), but it looks like there are additional blockers for this:
- `@metamask/eth-token-tracker@9.0.0` is on `@metamask/eth-block-tracker@10.0.0`
- `@metamask/network-controller@20.2.0` is on `@metamask/eth-block-tracker@9.0.3`

**Comment by MajorLift** [comment]
@bschorchit Thanks for checking in on this. We're just waiting on review and approval especially from codeowners. 

I just bumped this to priority-2 on the PR review queue, and I'll make a post today requesting another round of reviews.

**Review by gantunesr — DISMISSED** [review]
Approved for team-accounts.

Files reviewed,

- `app/scripts/lib/accounts/BalancesController.ts`
- `app/scripts/lib/snap-keyring/snap-keyring.test.ts`

File: app/scripts/snaps/preinstalled-snaps.ts
Comment (hmalik88) [inline app/scripts/snaps/preinstalled-snaps.ts:8]:
```suggestion
const PREINSTALLED_SNAPS = Object.freeze<ReadOnlyArray<PreinstalledSnap>>([
```
Can we do this instead since before the type was readonly?

Comment (MajorLift) [reply app/scripts/snaps/preinstalled-snaps.ts:8]: This would be redundant since the return type of `Object.freeze<T>` is `readonly T`. By hovering over `PREINSTALLED_SNAPS`, you can verify that its type is `readonly PreinstalledSnap[]` like before.

If we want to be extra careful about validating the type maybe we could add a type annotation or a `satisfies` clause?

```diff
diff --git a/app/scripts/snaps/preinstalled-snaps.ts b/app/scripts/snaps/preinstalled-snaps.ts
index 6be0bb7f35..95b9c54e69 100644
--- a/app/scripts/snaps/preinstalled-snaps.ts
+++ b/app/scripts/snaps/preinstalled-snaps.ts
@@ -10,6 +10,6 @@ const PREINSTALLED_SNAPS = Object.freeze<PreinstalledSnap[]>([
   ///: BEGIN:ONLY_INCLUDE_IF(build-flask)
   BitcoinWalletSnap as unknown as PreinstalledSnap,
   ///: END:ONLY_INCLUDE_IF
-]);
+]) satisfies readonly PreinstalledSnap[];
 
 export default PREINSTALLED_SNAPS;
```

Comment (daniela2000111) [reply app/scripts/snaps/preinstalled-snaps.ts:8]: Bug report

Comment (MajorLift) [reply app/scripts/snaps/preinstalled-snaps.ts:8]: @daniela2000111 Is there a bug related to the changes in `preinstalled-snaps.ts`?

Comment (hmalik88) [reply app/scripts/snaps/preinstalled-snaps.ts:8]: Ah yes that's correct! Fine to leave as is :)

Comment (hmalik88) [reply app/scripts/snaps/preinstalled-snaps.ts:8]: Curious about this bug report ^

**Comment by MajorLift** [comment]
@metamaskbot update-policies

File: app/scripts/snaps/preinstalled-snaps.ts
Comment (MajorLift) [inline app/scripts/snaps/preinstalled-snaps.ts:10]: @k-g-j I removed an unnecessary `as unknown as` cast while merging develop into this branch.

File: app/scripts/snaps/preinstalled-snaps.ts
Comment (MajorLift) [inline app/scripts/snaps/preinstalled-snaps.ts:11]: @k-g-j I removed an `as unknown as` cast introduced in https://github.com/MetaMask/metamask-extension/pull/26402 while merging develop into this branch.

**Review by owencraston — APPROVED** [review]

**Review by sleepytanya — APPROVED** [review]

**Review by hmalik88 — APPROVED** [review]

**Pull request merged**
